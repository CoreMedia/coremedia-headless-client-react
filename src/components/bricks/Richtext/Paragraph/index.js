// @flow
import { P as BasicParagraph } from '../../../basic/Paragraph';

const P = BasicParagraph.extend`
  margin-top: 0.5em;
`;
P.displayName = 'P';

export { P };
