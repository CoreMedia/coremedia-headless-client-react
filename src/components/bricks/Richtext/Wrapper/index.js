import Wrapper from './Wrapper';
import PictureWrapper from './PictureWrapper';

export { Wrapper, PictureWrapper };
