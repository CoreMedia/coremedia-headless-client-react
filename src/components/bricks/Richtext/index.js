// @flow
import React from 'react';
import PropTypes from 'prop-types';
import HtmlToReact from 'html-to-react';

import { H1, H2, H3 } from './Heading';
import { P } from './Paragraph';
import { Wrapper, PictureWrapper } from './Wrapper';
import Picture from '../CMPicture';

type Props = {
  text: string,
  color?: string,
};

const Richtext = ({ text, color }: Props) => {
  const processNodeDefinitions = new HtmlToReact.ProcessNodeDefinitions(React);
  const processingInstructions = [
    {
      shouldProcessNode: function(node) {
        return node.name === 'h1';
      },
      processNode: function(node, children, index) {
        return <H1 key={index}>{children}</H1>;
      },
    },
    {
      shouldProcessNode: function(node) {
        return node.name === 'h2';
      },
      processNode: function(node, children, index) {
        return <H2 key={index}>{children}</H2>;
      },
    },
    {
      shouldProcessNode: function(node) {
        return node.name === 'h3';
      },
      processNode: function(node, children, index) {
        return <H3 key={index}>{children}</H3>;
      },
    },
    {
      shouldProcessNode: function(node) {
        return node.name === 'p';
      },
      processNode: function(node, children, index) {
        return <P key={index}>{children}</P>;
      },
    },
    {
      shouldProcessNode: function(node) {
        return node.name === 'img';
      },
      processNode: function(node, children, index) {
        const src = node.attribs['cms-src'];
        const alt = node.attribs['alt'];
        return (
          <PictureWrapper key={index}>
            <Picture link={src} ratio="landscape_ratio16x9" alt={alt} stretch />
          </PictureWrapper>
        );
      },
    },
    {
      // handle all remaining nodes with default
      shouldProcessNode: function(node) {
        return true;
      },
      processNode: processNodeDefinitions.processDefaultNode,
    },
  ];

  const HtmlToReactParser = HtmlToReact.Parser;
  const htmlToReactParser = new HtmlToReactParser();
  const isValidNode = () => {
    return true;
  };
  const reactComponent = htmlToReactParser.parseWithInstructions(
    text,
    isValidNode,
    processingInstructions
  );
  return <Wrapper>{reactComponent}</Wrapper>;
};

Richtext.propTypes = {
  text: PropTypes.string.isRequired,
  color: PropTypes.string,
};

export default Richtext;
