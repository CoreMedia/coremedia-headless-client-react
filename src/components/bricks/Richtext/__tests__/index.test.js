// @flow
import React from 'react';

import Richtext from '..';

describe('Richtext Component', () => {
  it('should render HTML parsed to React Components', () => {
    const text = `<h1>Heading 1</h1><h2>Heading 2</h2><h3>Heading 3</h3><p>Paragraph</p><img cms-src="coremedia:///image/2662/data" alt="Scrum" class="float--left"/><a href="#none" cms-href="coremedia:///page/home~2450">CoreMedia Labs</a><ul><li><a href="https://labs.coremedia.com/"><strong>Github project</strong></a><li><a href="https://labs.coremedia.com/wiki"><strong>Technical wiki</strong></a></li></ul>`;
    const wrapper = shallow(<Richtext text={text} color="blue" />);
    expect(wrapper).toMatchSnapshot();
  });
});
