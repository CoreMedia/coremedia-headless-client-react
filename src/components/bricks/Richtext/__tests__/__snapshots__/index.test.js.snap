// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Richtext Component should render HTML parsed to React Components 1`] = `
<Wrapper
  theme={
    Object {
      "breakpoints": Object {
        "desktop": 992,
        "large": 1280,
        "tablet": 768,
      },
      "typography": Object {
        "fontFamily": Object {
          "text": "Helvetica",
        },
      },
    }
  }
>
  <H1
    key="0"
    theme={
      Object {
        "breakpoints": Object {
          "desktop": 992,
          "large": 1280,
          "tablet": 768,
        },
        "typography": Object {
          "fontFamily": Object {
            "heading": "Helvetica",
          },
          "fontWeight": Object {
            "normal": 400,
          },
        },
      }
    }
  >
    Heading 1
  </H1>
  <H2
    key="1"
    theme={
      Object {
        "breakpoints": Object {
          "desktop": 992,
          "large": 1280,
          "tablet": 768,
        },
        "typography": Object {
          "fontFamily": Object {
            "heading": "Helvetica",
          },
          "fontWeight": Object {
            "normal": 400,
          },
        },
      }
    }
  >
    Heading 2
  </H2>
  <H3
    key="2"
    theme={
      Object {
        "breakpoints": Object {
          "desktop": 992,
          "large": 1280,
          "tablet": 768,
        },
        "typography": Object {
          "fontFamily": Object {
            "heading": "Helvetica",
          },
          "fontWeight": Object {
            "normal": 400,
          },
        },
      }
    }
  >
    Heading 3
  </H3>
  <P
    key="3"
    theme={
      Object {
        "breakpoints": Object {
          "desktop": 992,
          "large": 1280,
          "tablet": 768,
        },
        "typography": Object {
          "fontFamily": Object {
            "text": "Helvetica",
          },
        },
      }
    }
  >
    Paragraph
  </P>
  <PictureWrapper
    key="4"
  >
    <WithTheme(Picture)
      alt="Scrum"
      color="red"
      link="coremedia:///image/2662/data"
      ratio="landscape_ratio16x9"
    />
  </PictureWrapper>
  <a
    cms-href="coremedia:///page/home~2450"
    href="#none"
    key="5"
  >
    CoreMedia Labs
  </a>
  <ul
    key="6"
  >
    <li
      key="0"
    >
      <a
        href="https://labs.coremedia.com/"
        key="0"
      >
        <strong
          key="0"
        >
          Github project
        </strong>
      </a>
    </li>
    <li
      key="1"
    >
      <a
        href="https://labs.coremedia.com/wiki"
        key="0"
      >
        <strong
          key="0"
        >
          Technical wiki
        </strong>
      </a>
    </li>
  </ul>
</Wrapper>
`;
